doc/Notifier.3
'\"
'\" Copyright (c) 1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
.so man.macros
.TH Notifier 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.Notifier -- The event queue and notifier API.

.SH METHODS
.sp
synchronized void \fBdeleteEvents\fR(EventDeleter \fIdeleter\fR)
.sp
synchronized int \fBdoOneEvent\fR(int \fIflags\fR)
.sp
static synchronized Notifier \fBgetNotifierForThread\fR(Thread
\fIthread\fR)
.sp
synchronized void \fBpreserve\fR()
.sp
synchronized void \fBqueueEvent\fR(TclEvent \fIevent\fR, int \fIpos\fR)
.sp
synchronized void \fBrelease\fR()

.SH ARGUMENTS
.AS TclObject "table[]" .
.AP Thread thread
The thread in which the Notifier should run.
.AP EventDeleter deleter
The EventDeleter determines which event in the event queue should be
deleted.
.AP TclEvent event
The event to put into the event queue.
.AP int pos
The position at which the event should be queued.
.AP int flags
What types of events to service.
.BE

.SH DESCRIPTION

The \fBNotifier\fR is the lowest-level part of the Jacl event
system. It is used by higher-level event sources such as file,
JavaBean and timer events. The Notifier manages an event queue that
holds \fBTclEvent\fR objects.

The Jacl \fBNotifier\fR is designed to run in a multi-threaded
environment. Each \fBNotifier\fR instance is associated with a
\fBprimary thread\fR. Any thread can queue (or dequeue) events using
the \fBqueueEvent\fR (or \fBdeleteEvents\fR) call. However, only the
primary thread may process events in the queue using the
\fBdoOneEvent\fR call. Attempts to call \fBdoOneEvent\fR from a
non-primary thread will cause a \fBTclRuntimeError\fR.

This class doesn't have a public constructor and cannot be directly
instantiated. You can get to it only through
\fBNotifier.getNotifierForThread\fR or the \fBgetNotifier\fR method of
the \fBInterp\fR class.

.TP
\fBgetNotifierForThread\fR
This static method returns the \fBNotifier\fR instance whose primary
thread is the given thread. If such an instance does not yet exist, it
is created automatically.
.TP
\fBpreserve\fR, \fBrelease\fR
These two methods maintain a reference count to manage the lifetime of
a Notifier and its corresponding event queue. When a Notifier is
created inside the \fB\fBgetNotifierForThread\fR method, its reference
count is zero. A call to preserve increments the reference count by
one; a call to \fBrelease\fR decrements the reference count by
one. When the reference count goes from one to zero, the Notifier is
disposed -- its event queue is deleted and any left-over events in the
queue will not be processed.
.sp
If a Java module intends to use a Notifier returned by
\fBgetNotifierForThread\fR over one or more method calls, it should
call \fBpreserve\fR immediately after \fBgetNotifierForThread\fR
returns. The module should call \fBrelase\fR when it no longer needs
access to the notifier.
.sp
NOTE: the notifier returned by \fIinterp\fR.\fBgetNotifier\fR will remain
valid as long as the Interpreter is valid. Hence, if a module always
access a notifier via calls to \fIinterp\fR.\fBgetNotifier\fR, it needs not
use the notifier's \fBpreserve\fR and \fBrelease\fR methods.
.TP
\fBqueueEvent\fR
This method puts the \fIevent\fR into the event queue at the position
specified by the \fIpos\fR argument. The \fIpos\fR argument may be one
of the following:
.RS
.TP
\fBTCL.QUEUE_TAIL\fR
Add the event at the back of the queue, so that all other pending
events will be serviced first.  This is almost always the right place
for new events.
.TP
\fBTCL.QUEUE_HEAD\fR
Add the event at the front of the queue, so that it will be serviced
before all other queued events. 
.TP
\fBTCL.QUEUE_MARK\fR
Add the event at the front of the queue, unless there are other events
at the front whose position is \fBTCL.QUEUE_MARK\fR; if so, add the
new event just after all other \fBTCL.QUEUE_MARK\fR events.  This
value of position is used to insert an ordered sequence of events at
the front of the queue, such as a series of Enter and Leave events
synthesized during a grab or ungrab operation in Tk.
.RE
.TP
\fBdeleteEvents\fR
This method provides a way to selectively delete events in the event
queue. It calls \fIdeleter\fR.\fBdeleteEvent\fR for each event in the
queue and deletes those for which \fIdeleter\fR.\fBdeleteEvent\fR
returns 1. Events for which the \fIdeleter\fR returns 0 are left in the
queue.
.TP
\fBdoOneEvent\fR
This method is called in the context of the Notifier's primary thread
to dispatch events in the event queue. If events are found in the
event queue, the \fBprocessEvent\fR method of the first available
event is called. If no event are in the queue, \fBdoOneEvent\fR checks
if there are idle handlers. If there are any, it invokes all of them
and returns. Finally, if no events or idle callbacks have been found,
then \fBdoOneEvent\fR sleeps until an event occurs, calles
\fBprocessEvent\fR the of the event and returns.
.sp
If the \fIflags\fR argument to \fBdoOneEvent\fR is non-zero, it
restricts the kinds of events that will be processed by
\fBdoOneEvent\fR. Flags may be an OR-ed combination of any of the
following bits:
.RS
.TP
\fBTCL.WINDOW_EVENTS\fR
Process window system events.
.TP
\fBTCL.FILE_EVENTS\fR
Process file events.
.TP
\fBTCL.TIMER_EVENTS\fR
Process timer events.
.TP
\fBTCL.IDLE_EVENTS\fR
Process idle callbacks.
.TP
\fBTCL.ALL_EVENTS\fR
Process all kinds of events: equivalent to OR-ing together all of the
above flags or specifying none of them.
.TP
\fBTCL.DONT_WAIT\fR
Don't sleep: process only events that are ready at the time of the
call.
.PP
If any of the flags \fBTCL.WINDOW_EVENTS\fR, \fBTCL.FILE_EVENTS\fR,
\fBTCL.TIMER_EVENTS\fR, or \fBTCL.IDLE_EVENTS\fR is set, then the only
events that will be considered are those for which flags are set.
Setting none of these flags is equivalent to the value
\fBTCL.ALL_EVENTS\fR, which causes all event types to be processed.
If an application has additional event sources then additional flag
values may also be valid, depending on those event sources.
.sp
The \fBTCL.DONT_WAIT\fR flag causes \fBdoOneEvent\fR not to put the
process to sleep: it will check for events but if none are found then
it returns immediately with a return value of 0 to indicate that no
work was done. \fBdoOneEvent\fR will also return 0 without doing
anything if the only alternative is to block forever (this can happen,
for example, if flags is \fBTCL.IDLE_EVENTS\fR and there are no idle
callbacks pending, or if no event handlers or timer handlers exist).
.sp
\fBdoOneEvent\fR may be invoked recursively. For example, it is
possible to invoke \fBdoOneEvent\fR recursively from an event handler
called by \fBdoOneEvent\fR. This sort of operation is useful in some
modal situations, such as when a notification dialog has been popped
up and an application wishes to wait for the user to click a button in
the dialog before doing anything else.
.RE

.SH "EQUIVALENT C FUNCTIONS"
Tcl_QueueEvent, Tcl_DeleteEvents, Tcl_DoOneEvent

.SH "SEE ALSO"
TclEvent, TimerHandler, IdleHandler, EventDeleter, Interp

.SH KEYWORDS
event, notifier, event queue, event sources, file events, timer, idle


ChangeLog
2019-05-04  Fabrice Ducos <fabrice.ducos@gmail.com>

        * Support of latest versions of Tcl (at least Tcl 8.5
        and Tcl 8.6) and of modern JDK (10+);
        support of JDK 4- abandoned (use of auto-boxing);
        support of OSX systems added;
        details below

        * configuration (tcljava.m4, configure.in),
        and compilation (Makefile.in) templates updated for
        JDK 10+ (javah not available anymore).
        javah is still used when it is available (JDK 9-).
        * minor update of jtclsh.in (for OSX dynamic libraries)
        * many obsolescent constructs (mostly old-style boxing)
        removed in *.java under src/jacl/tcl/lang,
        src/tclblend/tcl/lang and src/tcljava/tcl/lang,
        in order to silence many warnings.
        (effect: support of JDK 4- abandoned)
        * src/native/javaInterp.c: 2010 implementation
        of Java_tcl_lang_Interp_initName() crashed on the
        example given at docs/website/getstart.html,
        because calling Tcl_GetNameOfExecutable
        before Tcl_FindExecutable gives rise to undefined
        behaviour, at least in latest versions of Tcl
        (crashes observed on several systems, at least with
        Tcl 8.6).
        Java_tcl_lang_Interp_initName() is now simplified and
        always calls Tcl_FindExecutable first.
        * all the occurences of interp->result have been
        replaced with Tcl_GetStringResult(interp).
        Accessing interp->result directly was deprecated in
        latest versions of the Tcl C API, and removed from
        Tcl 8.6 (another solution was to #define USE_INTERP_RESULT,
        but this would have been an easy-going, quick and dirty way)
        ref. https://www.tcl.tk/man/tcl/TclLib/Interp.htm

2019-05-01  Fabrice Ducos <fabrice.ducos@gmail.com>

        * tcljava resurrected
        * migration from a CVS to a GIT repository
        (with reposurgeon)

2010-02-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* tests/tcl/regexp.test: Add test case for
        escaped - character in regexp command.
        [Bug 2012389]

2010-02-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceAll):
	Fixup regsub -all logic so 5 more tests pass.
	The replaceAll() logic is broken because it depends
	on matching substrings instead of using an index
	into the larger string and Java's regexp library
	uses this to calculate line starts in multiline mode.

2010-02-18  Mo DeJong  <mdejong@users.sourceforge.net>

        Incorporate fix/tests from Tcl head on 2010-02-11,
        aka was Tcl [Bug 2826551, Patch 2948425].

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
        Fixup increment index on zero length match issue.
	* tests/tcl/regexp.test: Back port of new regexp tests,
        some tests marked with {endIndex} due to lack of impl.

2010-02-11  Mo DeJong  <mdejong@users.sourceforge.net>

	Patch 2946118

	* src/jacl/tcl/lang/ListCmd.java: Minor style reformat.
	* src/jacl/tcl/lang/Util.java (scanElement, convertElement):
	Remove crusty old code that treated backslash zero
	as the end of the line. That logic was removed from
	C Tcl long ago.
	* tests/inprogress/EventAdaptor.test: Fixup tests
	that depended on backslash zero quirks.
	* tests/tcl/list.test: Add test cases for backslash zero
	as argument to list command.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (match, _match,
        replaceFirst, replaceAll): Rewrite -start offset
        logic to correctly deal with odd edge cases.
	* tests/tcl/regexp.test: Add additional test cases
        that cover -start index edge cases.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Rewrite argument processing logic to avoid using
        IndexOutOfBoundsException to deal with an incorrect
        number of arguments.

2009-10-03  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fix regsub -all crash because offset was not getting
        advanced by 1 on a match of length zero.
        Fixup -start argument support with and without -all
        in the case where the offset was larger than the
        length of the string.
	* tests/tcl/regexp.test: Add additional test cases.
        Enable regsub test cases that were getting caught
        in an infinite loop.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        failing tests related to optional subgroups.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Minor paren fix, no function changes.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Rework
        logic related to match result groups.
	* tests/tcl/regexp.test: Add missing test cases for
        non-capturing groups and -inline results.

2009-09-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup use of MULTILINE flag.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for -linestop and -lineanchor arguments to regexp.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):  Fixup
        support for -linestop and -lineanchor arguments to regsub.
	* tests/tcl/regexp.test: Add missing test cases for
        -linestop and -lineanchor arguments.

2009-09-22  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup support for the empty string as the string
        argument to regsub by removing the MULTILINE flag when
        an empty string is detected.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for empty string as string argument.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Remove comments.
	* tests/tcl/regexp.test: Add additional test cases

src/tclblend/tcl/lang/Notifier.java
/*
 * Notifier.java --
 *
 *	Implements the Blend version of the Notifier class.
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: Notifier.java,v 1.1.1.1 1998/10/14 21:09:10 cvsadmin Exp $
 *
 */

package tcl.lang;

import java.util.*;

/*
 * Implements the Blend version of the Notifier class. The Notifier is
 * the lowest-level part of the event system. It is used by
 * higher-level event sources such as file, JavaBean and timer
 * events. The Notifier manages an event queue that holds TclEvent
 * objects.
 *
 * The Notifier is designed to run in a multi-threaded
 * environment. Each notifier instance is associated with a primary
 * thread. Any thread can queue (or dequeue) events using the
 * queueEvent (or deleteEvents) call. However, only the primary thread
 * may process events in the queue using the doOneEvent()
 * call. Attepmts to call doOneEvent from a non-primary thread will
 * cause a TclRuntimeError.
 *
 * This class does not have a public constructor and thus cannot be
 * instantiated. The only way to for a Tcl extension to get an
 * Notifier is to call Interp.getNotifier() (or
 * Notifier.getNotifierForThread), which returns the Notifier for that
 * interpreter (thread).
 */

public class Notifier {

private static Notifier globalNotifier;

// First pending event, or null if none.

private TclEvent firstEvent;

// Last pending event, or null if none.

private TclEvent lastEvent;

// Last high-priority event in queue, or null if none.

private TclEvent markerEvent;

// The primary thread of this notifier. Only this thread should process
// events from the event queue.

Thread primaryThread;

// Reference count of the notifier. It's used to tell when a notifier
// is no longer needed.

int refCount;

/*
 *----------------------------------------------------------------------
 *
 * Notifier --
 *
 *	Creates a Notifier instance.
 *
 * Side effects:
 *	Member fields are initialized.
 *
 *----------------------------------------------------------------------
 */

private
Notifier(
    Thread primaryTh)		// The primary thread for this Notifier.
{
    primaryThread = primaryTh;
    firstEvent        = null;
    lastEvent         = null;
    markerEvent       = null;
    refCount = 0;
    init();
}

/*
 *----------------------------------------------------------------------
 *
 * getNotifierForThread --
 *
 *	Get the notifier for this thread, creating the Notifier,
 *	when necessary. Note that this is a partial implementation
 *	that only supports a single notifier thread.
 *
 * Results:
 *	The Notifier for this thread.
 *
 * Side effects:
 *	The Notifier is created when necessary.
 *
 *----------------------------------------------------------------------
 */

public static synchronized Notifier
getNotifierForThread(
    Thread thread)		// The thread that owns this Notifier.
{
    if (globalNotifier == null) {
	globalNotifier = new Notifier(thread);
    } else if (globalNotifier.primaryThread != thread) {
	return null;
    } 
    return globalNotifier;
}

/*
 *----------------------------------------------------------------------
 *
 * preserve --
 *
 *	Increment the reference count of the notifier. The notifier will
 *	be kept in the notifierTable (and alive) as long as its reference
 *	count is greater than zero.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The refCount is incremented.
 *
 *----------------------------------------------------------------------
 */

public synchronized void
preserve()
{
    if (refCount < 0) {
	throw new TclRuntimeError("Attempting to preserve a freed Notifier");
    }
    ++refCount;
}

/*
 *----------------------------------------------------------------------
 *
 * release --
 *
 *	Decrement the reference count of the notifier. The notifier will
 *	be free when its refCount goes from one to zero.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The notifier may be removed from the notifierTable when its
 *	refCount reaches zero.
 *
 *----------------------------------------------------------------------
 */

public synchronized void
release()
{
    if ((refCount == 0) && (primaryThread != null)) {
	throw new TclRuntimeError(
		"Attempting to release a Notifier before it's preserved");
    }
    if (refCount <= 0) {
	throw new TclRuntimeError("Attempting to release a freed Notifier");
    }
    --refCount;
    if (refCount == 0) {
	primaryThread = null;
	globalNotifier = null;
	dispose();
    }
}

/*
 *----------------------------------------------------------------------
 *
 * queueEvent --
 * 
 *	Insert an event into the event queue at one of three
 *	positions: the head, the tail, or before a floating marker.
 *	Events inserted before the marker will be processed in
 *	first-in-first-out order, but before any events inserted at
 *	the tail of the queue.  Events inserted at the head of the
 *	queue will be processed in last-in-first-out order.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	If this method is invoked by a non-primary thread, the
 *	primaryThread of this Notifier will be notified about the new
 *	event.
 *
 *----------------------------------------------------------------------
 */

public synchronized void
queueEvent(
    TclEvent evt,		// The event to put in the queue.
    int position)		// One of TCL.QUEUE_TAIL,
				// TCL.QUEUE_HEAD or TCL.QUEUE_MARK.
{
    evt.notifier = this;

    if (position == TCL.QUEUE_TAIL) {
	// Append the event on the end of the queue.

	evt.next = null;

	if (firstEvent == null) {
	    firstEvent = evt;
	} else {
	    lastEvent.next = evt;
	}
	lastEvent = evt;
    } else if (position == TCL.QUEUE_HEAD) {
	// Push the event on the head of the queue.

	evt.next = firstEvent;
	if (firstEvent == null) {
	    lastEvent = evt;
	}
	firstEvent = evt;
    } else if (position == TCL.QUEUE_MARK) {
	// Insert the event after the current marker event and advance
	// the marker to the new event.

	if (markerEvent == null) {
	    evt.next = firstEvent;
	    firstEvent = evt;
	} else {
	    evt.next = markerEvent.next;
	    markerEvent.next = evt;
	}
	markerEvent = evt;
	if (evt.next == null) {
	    lastEvent = evt;
	}
    } else {
	// Wrong flag.

	throw new TclRuntimeError("wrong position \"" + position +
	       "\", must be TCL.QUEUE_HEAD, TCL.QUEUE_TAIL or TCL.QUEUE_MARK");
    }

    if (Thread.currentThread() != primaryThread) {
	alertNotifier();
    }
}

/*
 *----------------------------------------------------------------------
 *
 * deleteEvents --
 *
 *	Calls an EventDeleter for each event in the queue and deletes
 *	those for which deleter.deleteEvent() returns 1. Events
 *	for which the deleter returns 0 are left in the queue.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Potentially removes one or more events from the event queue.
 *
 *----------------------------------------------------------------------
 */

public synchronized void
deleteEvents(
    EventDeleter deleter)	// The deleter that checks whether an event
				// should be removed.
{
    TclEvent evt, prev;

    for (prev = null, evt = firstEvent; evt != null; evt = evt.next) {
        if (deleter.deleteEvent(evt) == 1) {
            if (firstEvent == evt) {
                firstEvent = evt.next;
                if (evt.next == null) {
                    lastEvent = null;
                }
            } else {
                prev.next = evt.next;
            }
	    if (markerEvent == evt) {
		markerEvent = null;
	    }
        } else {
            prev = evt;
        }
    }
}

/*
 *----------------------------------------------------------------------
 *
 * serviceEvent --
 *
 *	Process one event from the event queue.
 *
 * Results:
 *	The return value is 1 if the procedure actually found an event
 *	to process. If no processing occurred, then 0 is returned.
 *
 * Side effects:
 *	Invokes all of the event handlers for the highest priority
 *	event in the event queue.  May collapse some events into a
 *	single event or discard stale events.
 *
 *----------------------------------------------------------------------
 */

synchronized int
serviceEvent(
    int flags)			// Indicates what events should be processed.
				// May be any combination of TCL.WINDOW_EVENTS
				// TCL.FILE_EVENTS, TCL.TIMER_EVENTS, or other
				// flags defined elsewhere.  Events not
				// matching this will be skipped for processing
				// later.
{
    TclEvent evt, prev;

    // No event flags is equivalent to TCL_ALL_EVENTS.
    
    if ((flags & TCL.ALL_EVENTS) == 0) {
	flags |= TCL.ALL_EVENTS;
    }

    // Loop through all the events in the queue until we find one
    // that can actually be handled.

    for (evt = firstEvent; evt != null; evt = evt.next) {
	// Call the handler for the event.  If it actually handles the
	// event then free the storage for the event.  There are two
	// tricky things here, both stemming from the fact that the event
	// code may be re-entered while servicing the event:
	//
	// 1. Set the "isProcessing" field to true. This is a signal to
	//    ourselves that we shouldn't reexecute the handler if the
	//    event loop is re-entered.
	// 2. When freeing the event, must search the queue again from the
	//    front to find it.  This is because the event queue could
	//    change almost arbitrarily while handling the event, so we
	//    can't depend on pointers found now still being valid when
	//    the handler returns.

	boolean b = evt.isProcessing;
	evt.isProcessing = true;

	if ((b == false) && (evt.processEvent(flags) != 0)) {
	    evt.isProcessed = true;
	    if (evt.needsNotify) {
		synchronized(evt) {
		    evt.notifyAll();
		}
	    }
	    if (firstEvent == evt) {
		firstEvent = evt.next;
		if (evt.next == null) {
		    lastEvent = null;
		}
		if (markerEvent == evt) {
		    markerEvent = null;
		}
	    } else {
		for (prev = firstEvent; prev.next != evt; prev = prev.next) {
		    // Empty loop body.
		}
		prev.next = evt.next;
		if (evt.next == null) {
		    lastEvent = prev;
		}
		if (markerEvent == evt) {
		    markerEvent = prev;
		}
	    }
	    return 1;
	} else {
	    // The event wasn't actually handled, so we have to
	    // restore the isProcessing field to allow the event to be
	    // attempted again.

	    evt.isProcessing = b;
	}

	// The handler for this event asked to defer it.  Just go on to
	// the next event.

	continue;
    }
    return 0;
}

/*
 *----------------------------------------------------------------------
 *
 * doOneEvent --
 *
 *	Process a single event of some sort.  If there's no work to
 *	do, wait for an event to occur, then process it. May delay
 *	execution of process while waiting for an event, unless
 *	TCL.DONT_WAIT is set in the flags argument.  This routine
 *	should only be called from the primary thread for the
 *	notifier.
 *	
 * Results:
 *	The return value is 1 if the procedure actually found an event
 *	to process. If no processing occurred, then 0 is returned
 *	(this can happen if the TCL.DONT_WAIT flag is set or if there
 *	are no event handlers to wait for in the set specified by
 *	flags).
 *
 * Side effects:
 *	May delay execution of process while waiting for an event,
 *	unless TCL.DONT_WAIT is set in the flags argument. Event
 *	sources are invoked to check for and queue events. Event
 *	handlers may produce arbitrary side effects.
 *
 *----------------------------------------------------------------------
 */

public native int
doOneEvent(
    int flags);			// Miscellaneous flag values: may be any
				// combination of TCL.DONT_WAIT,
				// TCL.WINDOW_EVENTS, TCL.FILE_EVENTS,
				// TCL.TIMER_EVENTS, TCL.IDLE_EVENTS,
				// or others defined by event sources.

/*
 *----------------------------------------------------------------------
 *
 * alertNotifier --
 *
 *	Wake up the C notifier.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	May cause the C level event loop to wake up.
 *
 *----------------------------------------------------------------------
 */

private final native void
alertNotifier();

/*
 *----------------------------------------------------------------------
 *
 * init --
 *
 *	Initialize the C event source for the Java notifier.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

private final native void
init();

/*
 *----------------------------------------------------------------------
 *
 * dispose --
 *
 *	Tear down the C event source for the Java notifier.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

final native void
dispose();

/*
 *----------------------------------------------------------------------
 *
 * hasEvents --
 *
 *	Check to see if there are events waiting to be processed.
 *
 * Results:
 *	Returns true if there are events on the Notifier queue.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

private final synchronized boolean
hasEvents()
{
    return (firstEvent != null);
}

} // end Notifier
